### PR TITLE
Resolve #1165, Resolve #1164, Resolve #1163, Resolve #1162, Resolve #1161, Resolve #1095, Resolve #879, Progress #1059 -- [QOL] Scripting

### LINKED ISSUE #879: Remove timer from logic code
_No content_

---
Beep, boop, I'm [a bot](https://github.com/LeagueSandbox/IssueBot)! This issue was created by `@Neekhaulas` in `#development`.

### LINKED ISSUE #1095: Minions Attack Nexus Before Turrets Are Destroyed
Minions seem to ignore the protection system and attack even when they shouldn't.

### LINKED ISSUE #1161: [Scripting] Optional Auto Face Direction
At the moment spell casts always force the caster to face the end position of the spell they cast. This should be optional in ScriptMetadata.

### LINKED ISSUE #1162: Targetability is ignored
Server-side targeting ignores the targetability of units. This means normally untargetable minions such as Zed shadows will still be targeted by turrets or via attack-move.

### LINKED ISSUE #1163: Non-existant buff script crash
Attempting to add a buff which does not have a script associated with it results in a crash. To prevent this, an empty buff script should be used as a default.

### LINKED ISSUE #1164: [Scripting] Expose Buff Script
Buff scripts are current private, and thus any script functionality which must share information between buff scripts and spell scripts is closed. The simplest fix is to make the `private IBuffGameScript _buffGameScript` into `public IBuffGameScript BuffScript { get; private set; }` in Buff.cs.

### LINKED ISSUE #1165: [Content] Buff Script Namespaces
All of the buff scripts use their class name as the namespace, which can lead to style issues when trying to reference the class name. The recommended change is to simply use `Buffs` as the namespace, similar to how `Spells` is used for spell scripts.

### PR DESCRIPTION
**Pull request by lzardy**
* Scripting:
  * Namespaces of all buff scripts changed to "Buffs".
  * Implemented Zed W
  * Fixed a crash with empty buff scripts.
  * Added AutoFaceDirection to script metadata.
  * Added OnSpellMissileEnd event.
  * AddParticle[Target] supports FXFlags.
  * AddMinion supports spawn variables such as IgnoreCollision, IsTargetable[ToTeamFlags], SkinID, and VisibilityOwner.
  * Added SetAutocast function.
  * Added CreateDeathData function for usage in IAttackableUnit.Die() function.
  * Added OverrideEndPosition to MissileParameters for usage with skillshots that are not fired from facing direction.
  * CreateTimer is marked as Obsolete.
* Minion:
  * Added IgnoreCollision, IsTargetable, and VisibilityOwner.
  * Reordered variables to align with SpawnMinionS2C packet.
  * Targeting accounts for targetable status.
* BaseTurret:
  * Targeting accounts for targetable status.
* Spell:
  * Fixed client-side crash with missiles being destroyed before spawning.
  * Added SetAutocast function.
* ObjAiBase:
  * SetTargetUnit does not accept untargetable or dead units.
  * Added SkinID variable and removed duplicates in higher level classes.
* Buff:
  * Script instance is now public.
  * Fixed a bug where removal status would be negated for buffs which re-activate.
* AttackableUnit:
  * Replication is now centralized to AttackableUnit.Update(), removing any other cases in higher level classes.
  * Death occurs in Update after Replication.Update(), allowing for stat mods from scripts to take effect before death.
  * Utilized DeathData for Die() function.
  * Collisions ignore ghosted units.
  * Targetable status is set upon initialization.
  * Removed SetIsTargetable in favor of SetStatus.
* Particle:
  * Added Flags variable.
* ObjectManager:
  * Missiles are destroyed if their target becomes untargetable.
  * Units targeting an untargetable unit have their target unset.
* Core:
  * Added DeathData.
  * Enums:
    * Added (incomplete) FXFlags.
* Packets:
  * Added NPC_Die_Broadcast, NPC_SetAutocast, and S2C_NPC_Die_MapView.
  
Resolve #1165
Resolve #1164
Resolve #1163
Resolve #1162
Resolve #1161
Resolve #1095
Resolve #879
Progress #1059

### CODE REVIEW

File: GameServerCore/Domain/ICharData.cs [hunk 1]
```diff
@@ -1,4 +1,5 @@
-﻿using GameServerCore.Enums;
+﻿using GameServerCore.Domain;
+using GameServerCore.Enums;
 
 namespace LeagueSandbox.GameServer.Content
```
Comment (danil179) [inline GameServerCore/Domain/ICharData.cs]: I assume that you should change the namespace here to `GameServerCore.Domain` as in the other files?

File: GameServerLib/GameObjects/AttackableUnits/AI/BaseTurret.cs [hunk 2]
```diff
@@ -200,7 +202,6 @@ public override void Update(float diff)
             }
 
             base.Update(diff);
-            Replication.Update();
```
Comment (danil179) [inline GameServerLib/GameObjects/AttackableUnits/AI/BaseTurret.cs:203]: Why replication is not updated here? I see that you do all the replications update in IAttackableUnit. I'll have to check if that works, as I'm not sure where or when replication should be updated.

File: GameServerLib/GameObjects/AttackableUnits/AttackableUnit.cs [hunk 3]
```diff
@@ -282,7 +298,7 @@ public override void OnCollision(IGameObject collider, bool isTerrain = false)
         /// <returns>True/False.</returns>
         public bool GetIsTargetableToTeam(TeamId team)
         {
-            if (!Stats.IsTargetable)
+            if (Status.HasFlag(StatusFlags.Targetable))
```
Comment (danil179) [inline GameServerLib/GameObjects/AttackableUnits/AttackableUnit.cs]: Is it missing a logical `not` here?

File: GameServerLib/GameObjects/AttackableUnits/AttackableUnit.cs [hunk 4]
```diff
@@ -457,8 +473,11 @@ public void RemoveStatModifier(IStatsModifier statModifier)
                 attackerId = (int)_game.PlayerManager.GetClientInfoByChampion((IChampion)attackerMinion.Owner).PlayerId;
             }
 
-            _game.PacketNotifier.NotifyUnitApplyDamage(attacker, this, damage, type, damageText,
-                _game.Config.IsDamageTextGlobal, attackerId, targetId);
+            if (attacker.Team != Team)
```
Comment (danil179) [inline GameServerLib/GameObjects/AttackableUnits/AttackableUnit.cs:476]: Do we have a flag `IsTargetableToTeam` or similar for this case? Or is it always true that you cannot attack your own team players?

**Review by danil179 — COMMENTED** [review]
Looks good to me. I'll have to test it as there are many changes. Also, I have a few comments.

File: GameServerLib/GameObjects/AttackableUnits/AttackableUnit.cs (hunk 4 above)
Comment (lzardy) [reply GameServerLib/GameObjects/AttackableUnits/AttackableUnit.cs:476]: This is just for the damage text. Minions often suicide using TakeDamage, so this prevents the text from appearing.

File: GameServerLib/GameObjects/AttackableUnits/AI/BaseTurret.cs (hunk 2 above)
Comment (lzardy) [reply GameServerLib/GameObjects/AttackableUnits/AI/BaseTurret.cs:203]: It should work fine as the Replication is overridden with the proper type, and each Replication update occurs in Update().

**Review by danil179 — APPROVED** [review]
Approved.

**Pull request merged**